internal/pkg/gateway/registry.go

```diff
@@ -260,68 +262,55 @@ func contains(slice []string, entry string) bool {
 
 // Returns a set of broadcastClients that can order a transaction for the given channel.
 func (reg *registry) orderers(channel string) ([]*orderer, error) {
-	err := reg.registerChannel(channel)
-	if err != nil {
-		return nil, err
-	}
 	var orderers []*orderer
-
-	// Get the config
-	config, err := reg.discovery.Config(channel)
-	if err != nil {
-		return nil, err
+	var ordererEndpoints []*endpointConfig
+	var err error
+	addr, exists := reg.channelOrderers.Load(channel)
+	// if it doesn't exist, get the orderers config for this channel
+	if exists {
+		ordererEndpoints = addr.([]*endpointConfig)
+	} else {
+		// no entry in the map - get the orderer config from discovery
+		ordererEndpoints, err = reg.updateConfig(channel)
+		if err != nil {
+			return nil, err
+		}
 	}
-
-	reg.configLock.RLock()
-	defer reg.configLock.RUnlock()
-
-	for _, eps := range config.GetOrderers() {
-		for _, ep := range eps.Endpoint {
-			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
-			if ord, ok := reg.broadcastClients[url]; ok {
-				orderers = append(orderers, ord)
+	for _, ep := range ordererEndpoints {
+		var client *orderer
+		entry, exists := reg.broadcastClients.Load(ep.address)
+		if exists {
+			client = entry.(*orderer)
+		} else {
+			// this orderer is new - connect to it and add to the broadcastClients registry
+			var tlsRootCerts [][]byte
+			// note the msp config is updated at the same time as the orderer config, so the following should be up to date (and nothing deletes from this map - yet)
+			cert, ok := reg.tlsRootCerts.Load(ep.mspid) // if this is not found, it's because it's a non-TLS connection
+			if ok {
+				tlsRootCerts = cert.([][]byte)
 			}
+			client, err = reg.endpointFactory.newOrderer(ep.address, ep.mspid, tlsRootCerts)
+			if err != nil {
+				// Failed to connect to this orderer for some reason.  Log the problem and skip to the next one.
+				reg.logger.Warnw("Failed to connect to orderer", "address", ep.address, "err", err)
+				continue
+			}
+			reg.broadcastClients.Store(ep.address, client)
```

Comment:
You need to lock. Otherwise two goroutines, each invoking **Submit()** can race each other to this point, and they will both create a new gRPC connection, and only one of them will be accessible in the map, and the other one will never be garbage collected and cause a resource leak.

Reply:
Yep.  Rather than add a broader lock, I have followed the pattern described here https://github.com/golang/go/issues/44159#issuecomment-780774977